### PR TITLE
q_date_in_lifetime

### PR DESCRIPTION
**Pull request by Luke-Sikina**
- Create query for validating whether a resource associated patient occurs during the patient's lifespan
- Create tests for query

### CODE REVIEW

File: cumulus_library_data_metrics/manifest.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ file_names = [
     "q_ref_target_valid/q_ref_target_valid.py",
     "q_system_use/q_system_use.py",
     "q_valid_us_core_v4/q_valid_us_core_v4.py",
+    "q_date_in_lifetime/q_date_in_lifetime.py",
```
Comment (mikix) [inline cumulus_library_data_metrics/manifest.toml]: Treat yourself to alphabet-sorting this into the list.

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE PATIENT_LIFESPAN AS
+SELECT
+    concat('Patient/', id) AS ref,
+    from_iso8601_timestamp(birthDate) - INTERVAL 30 DAY AS patient_start,
+    COALESCE(from_iso8601_timestamp(deceasedDateTime), CURRENT_TIMESTAMP) AS patient_end
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql]: This CURRENT_TIMESTAMP logic is not in the qualifier spec, yeah? That kind of date check is already handled in q_date_recent.

And we should be ignoring patients with just `deceasedBoolean=true` here (and thus no date) - this is currently treating those folks like living patients. I think all around, we should probably drop the "living" logic and this table should just persist a null `deceasedDateTime` into patient_end and just skip that check later if that's null. I think that would also be closer to the qualifier logic.

(And similarly, skip the birth date check later if it's null)

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        '{{ field }}' AS field,
+        {{ field }}  AS value,
+        (
+            from_iso8601_timestamp({{ field }}) > P.patient_start
+            AND from_iso8601_timestamp({{ field }}) < P.patient_end
+        ) AS valid
+    FROM {{ src }}
+        INNER JOIN PATIENT_LIFESPAN P ON P.ref = {{ src }}.{{ patient_field }}.reference
+    WHERE {{ field }} IS NOT NULL
+    {% if not loop.last %}UNION{% endif %}
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: 1. For performance, use UNION ALL when you don't need de-duplication, like here
2. ~Unioning here might mess up the stats - you could get multiple hits for one resource and have it be represented multiple times in the numerator. `q_date_resource` solved this by having the loop over date fields create a column for each entry.~ Looking at the code in base.summary.jinja, looks like we do a `distinct(id)` when calculating the numerator, so a UNION here is fine actually.

File: cumulus_library_data_metrics/q_date_in_lifetime/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+Any resource `R` that has a date field `R.date` and a patient reference `P` with a
+valid birthdate, `P.birthDate`, and, optionally, a date of death, `P.deceasedDateTime`.
+`R.date` must appear within the lifespan of the patient. For deceased patients, this is
+the period between `P.birthDate - 30 days`and `P.deceasedDateTime`. For living patients, this is
+the period between `P.birthDate - 30 days` and `CURRENT_TIMESTAMP`
+
+### Denominator
+
+Any resource `R` that has a date field `R.date` and a patient reference `P` with a
+valid birthdate, `P.birthDate`, and, optionally, a date of death, `P.deceasedDateTime`.
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/README.md]: I believe unless you define a different denominator in the python code, the denominator will just be all resources (like, all Procedures or all Conditions). Which I think is fine here - so maybe just update this document to say something like:
```
All possible rows for the resource in question.

(e.g. `count(*) from condition`)
```

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 5]
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE {{ study_prefix }}__q_date_in_lifetime_{{ src|lower }} AS
+{% for field in dates %}
+    SELECT
+        {{ src }}.{{ patient_field }}.reference as patient_ref,
+        id as resource_ref,
+        '{{ field }}' AS field,
+        {{ field }}  AS value,
+        (
+            from_iso8601_timestamp({{ field }}) > P.patient_start
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: This line is missing the annoying "partial date handling" rule about "latest possible date".

The thing that is a big bummer about that rule is it's gonna be super annoying to implement - we have a `from_iso8601_timestamp` UDF for duckdb, where it would be easy to stuff the logic. But that doesn't solve Athena. So you'll probably need to redefine the UDF in SQL itself as a Jinja macro and use that here?

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql [hunk 6]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE PATIENT_LIFESPAN AS
+SELECT
+    concat('Patient/', id) AS ref,
+    from_iso8601_timestamp(birthDate) - INTERVAL 30 DAY AS patient_start,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql]: The qualifier spec actually only defines a grace period of 30 days for the deceased date, and NOT the birth date. Which makes sense - some docs might trickle into the system after death. But probably not for the child before birth. Those would be linked to the parent's patient ID, I would assume.

**Review by mikix — COMMENTED** [review]
Didn't dig into the test edge cases here, just looked at the code bits for now

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja (hunk 5 above)
Comment (Luke-Sikina) [reply cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: make the date logic super complicated.

File: tests/data/q_date_in_lifetime/general/condition/weird_dates.ndjson [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{"resourceType": "Condition", "recordedDate": "2000", "id": "recorded-valid-year", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "Condition", "recordedDate": "2000-01", "id": "recorded-valid-year-month", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "Condition", "recordedDate": "2000-01-01T00:00:00", "id": "recorded-valid-time", "subject": {"reference": "Patient/dead"}}
```
Comment (Luke-Sikina) [inline tests/data/q_date_in_lifetime/general/condition/weird_dates.ndjson]: add tests for making birth date earliest date.

File: tests/test_metrics.py [hunk 8]
```diff
@@ -208,7 +211,7 @@ def run_study(
 
             args = [
                 "build",
-                # "--verbose",
+                "--verbose",
```
Comment (Luke-Sikina) [inline tests/test_metrics.py]: ahh, revert before mergnign.

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql [hunk 9]
```diff
@@ -0,0 +1,33 @@
+-- partial_date: iso date string
+-- suffix: to append to year if date is incomplete. Ex : '-01-01T23:59:59'
+CREATE MACRO DATE_WITH_FALLBACK_SUFFIX(partial_date, suffix) AS (
+    CASE -- Any date of format YYYY-MM-DD or more detailed, just parse
+        WHEN LENGTH(partial_date) > 9 THEN from_iso8601_timestamp(partial_date)
+        -- If there is a parsable year, capture that year with the suffix appended
+        WHEN regexp_matches(partial_date, '^[0-9]{4}', 'c') THEN from_iso8601_timestamp(concat(regexp_extract(partial_date, '^[0-9]{4}'), suffix))
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql]: Per [the spec](https://hl7.org/fhir/R4/datatypes.html#datetime): `The format is YYYY, YYYY-MM, YYYY-MM-DD or YYYY-MM-DDThh:mm:ss+zz:zz, e.g. 2018, 1973-06, 1905-08-23, 2015-02-07T13:28:17-05:00 or 2017-01-01T00:00:00.000Z`

So you need to handle YYYY-MM too. And since weird stuff like YYYY-M isn't allowed (doesn't mean it isn't in data, but... we probably don't need to be nice about parsing it)... you could probably get away with just a LENGTH check and simple concat rather than extracting.

Like:
```sql
...
WHEN LENGTH(partial_date) = 7 THEN from_iso(concat(partial_date, month_suffix))
WHEN LENGTH(partial_date) = 4 THEN from_iso(concat(partial_date, year_suffix))
ELSE from_iso(partial_date)
```

This does require using different suffixes, so it's a little more complicated. Especially since each month has a different number of days. Probably easier there to do a "less than next month's first day" situation. Unfortunately, the spec is not kind to us here.

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        -- Bail, take parser's best attempt, assume null
+        ELSE from_iso8601_timestamp(partial_date)
+    END
+);
+
+CREATE MACRO DATE_WITHIN_LIFESPAN(observation_date, birth, death) AS (
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql]: Using direct SQL macros is a little unusual for our code base - we typically lean on jinja for this. I'm not against it, but do you have a sense for how "standardized" the syntax is? We currently only support two dialects: Athena (aka Trino) and DuckDB. We might add another if we ever support Azure's Athena competitor. But do you know if this syntax works in Trino at least?

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        -- when the patient has a date of death, verify two things:
+        -- 1. latest possible interpretation of obs.date is after birth
+        -- 2. earliest possible interpretation of obs.date is before death
+        WHEN P.patient_end IS NOT NULL THEN (
+            DATE_WITH_FALLBACK_SUFFIX(observation_date, '-12-31T23:59:59') > birth AND
+            DATE_WITH_FALLBACK_SUFFIX(observation_date, '-01-01T00:00:00') < death
+        )
+        -- when the patient just has a birth date, verify just:
+        -- 1. latest possible interpretation of obs.date is after birth
+        ELSE DATE_WITH_FALLBACK_SUFFIX(observation_date, '-12-31T23:59:59') > birth
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql]: Birth date can be null too. So I think the logic here might be simpler if we broke it up into two separate checks? Something like this pseudo code:

```sql
(death IS NULL OR date < death)
AND
(birth IS NULL OR date > birth)
```

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    name = "q_date_in_lifetime"
+
+    def make_table(self, **kwargs) -> str:
+        self.add_summary_entry(kwargs["src"], self.render_sql("summary", **kwargs))
+        self.queries.append(self.render_sql(self.name, **kwargs))
+
+    def gen_patients_table(self):
+        path = os.path.dirname(__file__)
+        with open(f"{path}/patient_lifespan.sql") as file:
+            self.queries.append(file.read())
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py]: nit: this feels like it could be a utility function on `MetricMixin` (if it took a filename fragment). Like `add_sql()` or some other name that makes sense with its sibling functions.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 13]
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE {{ study_prefix }}__q_date_in_lifetime_{{ src|lower }} AS
+{% for field in dates %}
+    SELECT
+        {{ src }}.{{ patient_field }}.reference as patient_ref,
+        P.patient_start AS date_of_birth,
+        P.patient_end AS date_of_death,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: nit: why have two different names for these concepts? I'd vote to either keep patient_start here or change the SQL to use date_of_birth, vs having two names.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+        id as resource_ref,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: Caution around confusing IDs and References - IDs are "1234" and References are "Condition/1234" - so maybe rename this field as `resource_id` or do a `concat("{{ resource_type }}/", id) AS resource_ref`

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+        '{{ field }}' AS field,
+        {{ field }}  AS value,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: Trailing comma will fail in Athena I think.

That's a test one or both of us should do before landing this PR - running in Athena against our real data to confirm the syntax all works for a Trino engine.

File: cumulus_library_data_metrics/q_date_in_lifetime/README.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
+# q_date_in_lifetime
+
+**Are dates for resources that are associated with a patient within that patient's life?**
+
+Each patient has a `Patient.birthDate`, with many also having a `Patient.deceasedDateTime`.
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/README.md]: birthDate is optional too, sadly

File: cumulus_library_data_metrics/q_date_in_lifetime/README.md [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+Every observation with a date element should occur while the patient is alive, minus some margin (30 days).
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/README.md]: ~observation~ for "resource" I'd say - Observation is a type of FHIR resource, so the reader might think this metric is limited to those.

File: cumulus_library_data_metrics/q_date_in_lifetime/README.md [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+### Numerator (population)
+
+Any resource `R` that has a date field `R.date` and a patient reference `P` with a
+valid birthdate, `P.birthDate`, and, optionally, a date of death, `P.deceasedDateTime`.
+`R.date` must appear within the lifespan of the patient. For deceased patients, this is
+the period between `P.birthDate - 30 days`and `P.deceasedDateTime`. For living patients, this is
+the period between `P.birthDate - 30 days` and `CURRENT_TIMESTAMP`
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/README.md]: The 30 days is on the deceased side now and we don't check CURRENT

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+from cumulus_library_data_metrics.base import MetricMixin
+from cumulus_library_data_metrics.resource_info import DATES
+
+
+class TargetPopBuilder(MetricMixin, cumulus_library.BaseTableBuilder):
+    name = "q_date_in_lifetime"
+
+    def make_table(self, **kwargs) -> str:
+        self.add_summary_entry(kwargs["src"], self.render_sql("summary", **kwargs))
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py]: I think you're passing the denominator sql in as the `group` field - I think you want `denominator=self.render_sql()` - and maybe rename summary.sql to denominator.sql for clarity.

File: cumulus_library_data_metrics/q_date_in_lifetime/denominator.jinja [hunk 20]
```diff
@@ -0,0 +1,10 @@
+{% for field in dates %}
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/denominator.jinja]: OK is the intent here that you want to be doing numerator and denominators on FIELDS, not RESOURCES? (i.e. be able to say "there are 3 out of 15 date fields that are bad" rather than "there are 2 out of 5 resources with bad dates")

I'm not sure this will work like you want because I think base.summary.jinja is doing math on distinct `id` fields from the numerator and denominator. So you could do what you want, but add the field name to the id field maybe.

But I'd also just pause and ask - why not just operate on resources, rather than fields? The qualifier spec talks about this at the resource level. And you'd fight less with the data-metrics machinery, which is largely designed to think about things at the resource level. Like, you could drop this whole file I think in that case.

File: cumulus_library_data_metrics/manifest.toml [hunk 21]
```diff
@@ -12,11 +12,13 @@ file_names = [
     "c_resources_per_pt/c_resources_per_pt.py",
     "c_system_use/c_system_use.py",
     "c_us_core_v4_count/c_us_core_v4_count.py",
+    "q_date_in_lifetime/q_date_in_lifetime.py",
     "q_date_recent/q_date_recent.py",
     "q_ref_target_pop/q_ref_target_pop.py",
     "q_ref_target_valid/q_ref_target_valid.py",
     "q_system_use/q_system_use.py",
     "q_valid_us_core_v4/q_valid_us_core_v4.py",
+    "q_date_in_lifetime/q_date_in_lifetime.py",
```
Comment (mikix) [inline cumulus_library_data_metrics/manifest.toml]: ✂️

**Review by mikix — COMMENTED** [review]
Still bailed on looking at tests, just some more comments for now

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja (hunk 13 above)
Comment (Luke-Sikina) [reply cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja]: Yea, that's fair. start and end made more sense to me when I was using current date when dod was null. Guess I didn't switch all the way back.

File: cumulus_library_data_metrics/utils.jinja [hunk 22]
```diff
@@ -378,3 +378,30 @@ WHERE cnt >= {{ min_bucket_size }}
     )
 )
 {%- endmacro %}
+
+{% macro date_within_lifespan(observation_date, birth, death) -%}
```
Comment (mikix) [inline cumulus_library_data_metrics/utils.jinja]: nit: This seems so specific to this one metric, it might make more sense in the metric code, rather than here in the generic utilities.

File: cumulus_library_data_metrics/utils.jinja [hunk 23]
```diff
@@ -378,3 +378,30 @@ WHERE cnt >= {{ min_bucket_size }}
...
+(
+    ({{ death }} IS NULL OR {{ date_with_fallback_suffix(observation_date, '-01-01T00:00:00') }} < {{ death }})
+    AND
+    ({{ birth }} IS NULL OR {{ date_with_fallback_suffix(observation_date, '-12-31T23:59:59') }} > {{ birth }})
+)
+{%- endmacro %}
+
+-- partial_date: iso date string
+-- suffix: to append to year if date is incomplete. Ex : '-01-01T23:59:59'
+{% macro date_with_fallback_suffix(partial_date, suffix) -%}
```
Comment (mikix) [inline cumulus_library_data_metrics/utils.jinja:384]: This is implemented in an extremely SQL way, but I think it's good 👍 - all the very specific lengths are appropriate per the spec, and if we see non-spec/malformed data, it's fine to ignore. Comments are nice too.

File: cumulus_library_data_metrics/utils.jinja [hunk 24]
```diff
@@ -378,3 +378,30 @@ WHERE cnt >= {{ min_bucket_size }}
...
+-- date_with_fallback_suffix
+CASE -- Any date of format YYYY-MM-DD or more detailed, just parse
+    WHEN LENGTH({{ partial_date }}) > 9 THEN from_iso8601_timestamp({{ partial_date }})
```
Comment (mikix) [inline cumulus_library_data_metrics/utils.jinja]: I think this should be `> 10` yeah?

File: cumulus_library_data_metrics/utils.jinja [hunk 25]
```diff
@@ -378,3 +378,30 @@ WHERE cnt >= {{ min_bucket_size }}
     )
 )
 {%- endmacro %}
+
+{% macro date_within_lifespan(observation_date, birth, death) -%}
+-- date_within_lifespan
+(
+    ({{ death }} IS NULL OR {{ date_with_fallback_suffix(observation_date, '-01-01T00:00:00') }} < {{ death }})
```
Comment (mikix) [inline cumulus_library_data_metrics/utils.jinja]: nit: this one _could_ just use `from_iso8601_timestamp` as-is, since defaulting to earliest is how that's implemented. We only need the fallback suffix if we want a different time. But doesn't matter, and maybe the symmetry with the next line is nice.

But come to think of it, maybe a note about that would be nice in the comment above `date_with_fallback_suffix` -- like, how it's different from `from_iso8601_timestamp`

File: cumulus_library_data_metrics/utils.jinja [hunk 26]
```diff
@@ -378,3 +378,30 @@ WHERE cnt >= {{ min_bucket_size }}
...
+    AND
+    ({{ birth }} IS NULL OR {{ date_with_fallback_suffix(observation_date, '-12-31T23:59:59') }} > {{ birth }})
```
Comment (mikix) [inline cumulus_library_data_metrics/utils.jinja]: You don't want to throw a `.999` on there? What if someone is born at a few milliseconds to midnight?!?! j/k who cares, and I don't actually know how narrow SQL timestamps get.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+class TargetPopBuilder(MetricMixin, cumulus_library.BaseTableBuilder):
+    name = "q_date_in_lifetime"
+
+    def make_table(self, **kwargs) -> None:
+        self.add_summary_entry(kwargs["src"])
+        self.queries.append(self.render_sql(self.name, **kwargs))
+
+    def add_metric_queries(self) -> None:
+        self.queries.append(self.render_sql(template="patient_lifespan"))
+        for resource in DATES.keys():
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.py]: smallest possible nit: you don't need `.keys()`, that's the default iteration.

File: cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.jinja [hunk 28]
```diff
@@ -0,0 +1,8 @@
+{% import 'utils.jinja' as utils %}
+
+CREATE TABLE PATIENT_LIFESPAN AS
+SELECT
+    concat('Patient/', id) AS ref,
+    {{ utils.date_with_fallback_suffix('birthDate', '-01-01T00:00:00') }} AS date_of_birth,
+    {{ utils.date_with_fallback_suffix('deceasedDateTime', '-12-31T23:59:59') }} + INTERVAL 30 DAY AS date_of_death
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.jinja]: Looks like Trino (and thus Athena) supports the INTERVAL keyword, but just chiming in here to say that you should run your metric in Athena for realz before landing, to confirm it works as expected (at minimum, doesn't blow up - if you're feeling fancy, actually find an example of it working as expected).

And - generally, the INTERVAL 30 days thing works fine when wrapping around years and months yeah? I assume your tests cover that (haven't looked yet), and it would be crazy for the syntax not to work in that case - but I always get nervous when I see any sort of date handling.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+{% for field in dates %}
+    SELECT
+        {{ src }}.{{ patient_field }}.reference as patient_ref,
+        P.date_of_birth AS date_of_birth,
+        P.date_of_death AS date_of_death,
+        id as resource_id,
+        '{{ field }}' AS field,
+        {{ field }}  AS value
+    FROM {{ src }}
+        INNER JOIN PATIENT_LIFESPAN P ON P.ref = {{ src }}.{{ patient_field }}.reference
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja:34]: extremely minor nit: we don't generally capitalize table names or aliases, just keywords. It doesn't matter, just a tiny style difference.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 30]
```diff
@@ -0,0 +1,27 @@
+{% import 'utils.jinja' as utils %}
+
+CREATE TABLE {{ study_prefix }}__q_date_in_lifetime_{{ src|lower }} AS
+WITH invalid_fields AS (
+{% for field in dates %}
+    SELECT
+        {{ src }}.{{ patient_field }}.reference as patient_ref,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja:27]: Because you use `patient_field`, you probably want to add a check for that in the python code. Like where it's currently `for resource in DATES`, maybe add a check that the resource is also in `PATIENTS` (like if we later add support for `Practitioner` or similar, which isn't patient linked but has dates).

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+        P.date_of_birth AS date_of_birth,
+        P.date_of_death AS date_of_death,
+        id as resource_id,
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja:30]: Most of our other tables would put this as `id` (and less importantly, as the first column). I think the summary code even looks for the `id` field explicitly. Which makes me wonder about how this is getting summarized.

File: cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+    {% if not loop.last %}UNION ALL{% endif %}
+{% endfor %}
+)
+SELECT
+    patient_ref,
+    resource_id,
+    -- order by just to make test pass deterministically
+    ARRAY_AGG(field ORDER BY field DESC) as invalid_fields
+FROM invalid_fields
+GROUP BY patient_ref, resource_id;
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/q_date_in_lifetime.jinja:47]: I think you could just drop this extra SELECT and keep the intermediate union table above as the final product? The summary code already only cares about `DISTINCT id` so you don't need to do the aggregation yourself.

Also, it might be helpful for someone trying to debug a data issue to see the actual values that triggered this. (which would be fixed by dropping the SELECT but you could also add an array of values here too, if you wanted to keep this format.)

File: cumulus_library_data_metrics/q_date_in_lifetime/README.md [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+
+### Numerator (population)
+
+Given: any resource `R` that has a date field `R.date`, and a patient reference `P` with a
+valid birthdate `P.birthDate` or a date of death `P.deceasedDateTime`.
+`R.date` must appear within the lifespan of the patient, as defined below:
+- Deceased patients: the period between `P.birthDate`and `P.deceasedDateTime + 30 days`
+- Living patients: the period between `P.birthDate` and the end of time.
+- Deceased patients with no `P.birthDate`: the period between the beginning of time and
+`P.deceasedDateTime + 30 days`
```
Comment (mikix) [inline cumulus_library_data_metrics/q_date_in_lifetime/README.md:16]: We are well positioned to handle vampires in the dataset.

File: tests/data/q_date_in_lifetime/general/expected_allergyintolerance.csv [hunk 34]
```diff
@@ -0,0 +1,9 @@
+patient_ref,resource_id,invalid_fields
```
Comment (mikix) [inline tests/data/q_date_in_lifetime/general/expected_allergyintolerance.csv]: Just putting this comment somewhere - you're missing an `expected_summary.csv` file, which might be interesting to confirm that the summarization flow is working.

File: tests/data/q_date_in_lifetime/general/in_allergyintolerance.ndjson [hunk 35]
```diff
@@ -0,0 +1,16 @@
+{"resourceType": "AllergyIntolerance", "recordedDate": "2000-01-01", "id": "valid-recorded", "patient": {"reference": "Patient/dead"}}
```
Comment (mikix) [inline tests/data/q_date_in_lifetime/general/in_allergyintolerance.ndjson:1]: You went pretty hard on the test suite - maybe I should have poked my head in here earlier - as you may be able to see in other test suites, I have not been quite as exhaustive as you - which isn't bad on your part at all. Just maybe saying that my expectations were lower - like, I think it's reasonable to maybe have one resource (allergy intolerance, say) that you go ham on with all the weird edge conditions, then spot check a few other instances of other resource fields to just confirm that, yes you are hitting all the fields in DATES - but I myself would trust that DATES is well tested elsewhere, and we don't need to entirely re-test it here by checking every possible field.

But now that you already did it, it's nice. 😄  Though, if you do decide to change the format of your output tables, maybe that is a good opportunity to cut down on the volume here, since it's annoying to change all the expected tables. (You don't have to - you're being correctly rigorous, I'm just saying I'm not so rigorous myself.)

File: tests/data/q_date_in_lifetime/general/in_weird_dates.ndjson [hunk 36]
```diff
@@ -0,0 +1,9 @@
...
+{"resourceType": "Condition", "recordedDate": "2000-01", "id": "recorded-valid-year-month", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "Condition", "recordedDate": "2000-01-01T00:00:00", "id": "recorded-valid-time", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "Condition", "recordedDate": "2000", "id": "recorded-valid-same-as-dod", "subject": {"reference": "Patient/1day"}}
+{"resourceType": "Condition", "recordedDate": "1999", "id": "recorded-valid-same-as-dob", "subject": {"reference": "Patient/1day"}}
+{"resourceType": "Condition", "recordedDate": "2020-01-29", "id": "recorded-29d-after-dod", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "Condition", "recordedDate": "2020-02-01", "id": "invalid-recorded-31d-after-dod", "subject": {"reference": "Patient/dead"}}
+{"resourceType": "AllergyIntolerance", "onsetPeriod": {"start": "1776"}, "id": "recorded-valid-no-birth", "patient": {"reference": "Patient/undead"}}
+{"resourceType": "AllergyIntolerance", "onsetPeriod": {"start": "1776", "end": "9999"}, ",id": "recorded-valid-dates-at-all", "patient": {"reference": "Patient/beyond-mortal-comprehension"}}
```
Comment (mikix) [inline tests/data/q_date_in_lifetime/general/in_weird_dates.ndjson]: This is going to do something weird - I forget exactly what. I would... expect that the AllergyIntolerances are dropped on the floor? Which may be hard to see in the test suite, because these two examples happen to both be valid.

In general, it's invalid to have different resource types in the same file, per the FHIR spec.

File: cumulus_library_data_metrics/manifest.toml [hunk 37]
```diff
@@ -12,6 +12,7 @@ file_names = [
     "c_resources_per_pt/c_resources_per_pt.py",
```
Comment (mikix) [inline cumulus_library_data_metrics/manifest.toml:12]: You'll want to add `q_date_in_lifetime_summary` to the `flat_list` list in this same file. That way it will be exported from Athena, so it can be pushed up to the aggregator. (The summary is the main "result artifact" from the metric - the intermediate table that you actually wrote is mostly used for debugging (and to generate the summary.)

**Review by mikix — APPROVED** [review]
I stand by the two unresolved comments in the jinja file:
- it's worth checking `if resource in PATIENTS:` in the builder to avoid issues later when we add non-Patient-linked resources with dates like Practitioner
- you could drop the extra SELECT with its grouping to save some time and make debugging easier (this is going to be annoying for you, because it will drastically change the test files)

But neither are blockers, if you don't have time or inclination to touch those up. Thanks for all the running around on this one! :smile:
